### DISCUSSION TITLE
Handle ErrorResolution.Throw

### DISCUSSION THREAD

**Discussion by VagueGit** [Mimetis/Dotmim.Sync#835] [Q&A]
Playing with [ErrorResolution](https://dotmimsync.readthedocs.io/Errors.html#resolution) in DMS 0.97 and very excited. We currently handle sync errors with something like the following pseudo-code
```
try
{
  syncResult = await syncAgent.SynchronizeAsync(scopeName: scopeName, setup: syncSetup, syncType: syncType, progress: progress);
}
catch (HttpSyncWebException ex) when (blah blah blah) 
{
  logging and other stuff
}
catch (SyncException ex) when ( blah blah blah)
{ 
  logging and different stuff
}
etc
```
By default the local orchestrator as per the [docs for ErrorResolution.Throw](https://dotmimsync.readthedocs.io/Errors.html#errorresolution-throw) is configured to throw

It would be nice if in handling sync exceptions we could access `localOrchestrator.OnApplyChangesErrorOccured args.Exception.Message` and `args.ErrorRow` so they could be integrated with exception handling and logging.

Always receptive to advice on a better way of course.

**Comment by Mimetis** [comment]
I don't get it

You have all these information in the event

``` cs
        agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
        {
            Console.WriteLine($"ERROR: {args.Exception.Message}");
            Console.WriteLine($"ROW  : {args.ErrorRow}");
        });
```

Yo can log from there right ?

**Comment by VagueGit** [comment]
Thank you for the prompt response. That's true but we also do other stuff apart from logging and it would be nice if there was some way to maybe push `Exception.Message` and `ErrorRow` out with `ErrorResolution.Throw`. It might also give us options that we haven't really thought about.

**Reply by Mimetis** [reply]
thanks for the feedback.
I don't know if it's faisable, the general exception raised at the end may not be able to encapsualte the specific interceptor arguments errors, but I will look into it !

**Reply by VagueGit** [reply]
The other thing too is that if the error is logged in 
```
 agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
        {
            Console.WriteLine($"ERROR: {args.Exception.Message}");
            Console.WriteLine($"ROW  : {args.ErrorRow}");
        });
```
It then gets logged again when the error thrown is handled, so it would be nice not to log it here and only log it once where we do our sync error handling.

**Comment by VagueGit** [comment]
Pushing this idea a little further ... we return sync errors from the web api like this
`await webServerAgent.WriteExceptionAsync(this.Request, this.HttpContext.Response, ex, additionalInfo);`

It would be nice if `additionalInfo` were `Dictionery<string,string>` so we could pass back `args.ErrorRow`, `args.SchemaTable` etc or even better still change
`WriteExceptionAsync(HttpRequest httpRequest, HttpResponse httpResponse, Exception ex, string additionalInfo = null)`
to
`WriteExceptionAsync(HttpRequest httpRequest, HttpResponse httpResponse, Exception ex, ApplyChangesErrorOccuredArgs args)`

Stop me if you think I'm getting carried away :-)

**Comment by VagueGit** [comment]
Assuming we all want to see `ApplyChangesErrorOccuredArgs`, would it be possible for `ApplyChangesErrorOccuredArgs` to become a property of `SyncException`? Then we could access this property in our exception handlers without having to configure anything.

**Reply by Mimetis** [reply]
The problem is that general exception handling (the one you catch at the end) is basically for any exception.
I mean, you may have an exception that does not involve a row... like a connection exception for instance.
This exception raised by a bad connection cannot have a `ApplyChangesErrorOccuredArgs` 

That being said, your feedback is valuable, I guess we can try something, but it will require a little bit of refactoring and some adjustments in `SyncException`.

Will look into it and see what I can do !

**Reply by VagueGit** [reply]
`SyncException.ApplyChangesErrorOccuredArgs` could be nullable

**Reply by Mimetis** [reply]
I think I get a solution, when an error is raised by `OnApplyChangesErrorOccured`
You will have the complete exception in the inner exception.
It's in the main branch for now, and will be part of the next release

Thanks for the feedback !!!
